hail/python/test/hail/matrixtable/test_matrix_table_from_parts.py

```diff
@@ -0,0 +1,130 @@
+                   , **unless(no_col_props, {'bar':'c'})
+                   , **unless(no_entry_props, {'baz':3})
+                   )
+        , hl.Struct( row_idx=1
+                   , **unless(no_col_props, {'bar':'d'})
+                   , col_idx=1
+                   , **unless(no_row_props, {'foo':'b'})
+                   , **unless(no_entry_props, {'baz':4})
+                   )
+        ]
+
+def test_from_parts():
+    mt = hl.MatrixTable.from_parts( globals=Inits.globals
+                                  , rows=Inits.rows
+                                  , cols=Inits.cols
+                                  , entries=Inits.entries
+                                  )
+    assert_matches_globals(mt)
+    assert_matches_rows(mt)
+    assert_matches_cols(mt)
+    assert_matches_entries(mt)
+
+def test_optional_globals():
+    mt = hl.MatrixTable.from_parts( rows=Inits.rows
+                                  , cols=Inits.cols
+                                  , entries=Inits.entries
+                                  )
+    assert_matches_globals(mt, no_props=True)
+    assert_matches_rows(mt)
+    assert_matches_cols(mt)
+    assert_matches_entries(mt)
+
+def test_optional_rows():
+    mt = hl.MatrixTable.from_parts( globals=Inits.globals
+                                  , cols=Inits.cols
+                                  , entries=Inits.entries
+                                  )
+    assert_matches_globals(mt)
+    assert_matches_rows(mt, no_props=True)
+    assert_matches_cols(mt)
+    assert_matches_entries(mt, no_row_props=True)
+
+def test_optional_cols():
+    mt = hl.MatrixTable.from_parts( globals=Inits.globals
+                                  , rows=Inits.rows
+                                  , entries=Inits.entries
+                                  )
+    assert_matches_globals(mt)
+    assert_matches_rows(mt)
+    assert_matches_cols(mt, no_props=True)
+    assert_matches_entries(mt, no_col_props=True)
+
+def test_optional_globals_and_cols():
+    mt = hl.MatrixTable.from_parts( rows=Inits.rows
+                                  , entries=Inits.entries
+                                  )
+    assert_matches_globals(mt, no_props=True)
+    assert_matches_rows(mt)
+    assert_matches_cols(mt, no_props=True)
+    assert_matches_entries(mt, no_col_props=True)
+
+def test_optional_globals_and_rows_and_cols():
+    mt = hl.MatrixTable.from_parts(entries=Inits.entries)
+    assert_matches_globals(mt, no_props=True)
+    assert_matches_rows(mt, no_props=True)
+    assert_matches_cols(mt, no_props=True)
+    assert_matches_entries(mt, no_row_props=True, no_col_props=True)
+
+def test_optional_entries():
+    mt = hl.MatrixTable.from_parts(rows=Inits.rows, cols=Inits.cols)
+    assert_matches_globals(mt, no_props=True)
+    assert_matches_rows(mt)
+    assert_matches_cols(mt)
+    assert_matches_entries(mt, no_entry_props=True)
+
+def assert_raises_when_no_rows_and_entries():
+    with pytest.raises(AssertionError):
```

Comment:
Ah, nice, I wasn't aware of the `pytest.raises`. I like how concise this is compared to what I've done elsewhere:
```
try:
    ...
except AssertionError err:
    assert ... err.args[0] ...
else:
    assert False
```